### PR TITLE
feat: security improvements and features

### PR DESCRIPTION
**Pull request by StLyn4**
## Description

Hi all. Well... Let's start in order. Some of the changes were made to make the rest work.

All changes below are draft in nature. I mean, they're finished and fully operational, but I'd love to hear feedback and make edits.

- Hiding details of unhandled errors (with code 500, related to discussion #2669):
     - If the error does not have a status code, then it is automatically set to 500.
     - The error message will then be replaced with "Something went wrong." (unless `config.debug` is `true`)
     - Changes applied to both REST and GraphQL. Also, the `extensions.statusCode` field has been added to the responses of the second.
     - Full information will be displayed in the console, as before.
     - Added tests for changes above.
- Added `collectionConfig.auth.removeTokenFromResponses` flag:
     - When this flag is added, all built-in requests that send an authorization token (JWT) in the response body will have the token removed. Here I need to explain the situation a little. Let's say I decide to only use http-only authentication cookies to prevent an attack vector where an attacker's JS code can steal the token. In this case, I face the problem that such code can send a request to `/me` or `/refresh-token` at any time to get what it wants in the response. In such a case, the http-only flag loses its effect. If the user is sure that he will not use the token from the response body, then he can now apply the new flag.
     - Added tests for changes above.
- Changed [RESTClient helper](https://github.com/payloadcms/payload/blob/6125b66286e5315725ca0ae365c81a04c1c1a54c/test/helpers/rest.ts#L108). Here I would especially like to get feedback:
     - The code of methods has been redesigned to look more or less similar to each other.
     - In the `login` method, if there is no token in the response, then it will be taken from the cookies.
     - Added `logout` method.
     - By default, all requests pass authorization data (can be disabled via `auth: false`). Previously, there was different from method to method.
     - Added `endpointWithAuth` method, which, unlike `endpoint`, passes authorization data

- [x] I have read and understand the CONTRIBUTING.md document in this repository

## Type of change

- [x] _[Possible]_ Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist:

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Existing test suite passes locally with my changes
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by AlessioGr** [comment]
Hey @StLyn4 

Big thanks for your PR, much appreciated!

So, I was looking over the changes you suggested about error handling. Ever since we kicked passport-local-mongoose to the curb, it seems like the error messages are behaving pretty well now. Could you take a second look and let me know if you still see any issues? If things are cool, could you roll back the error handling part of your PR?

The rest of your changes look good! About hiding away the unhandled errors, I'm a bit concerned that hiding these could mess up things for someone who's relying on those error messages. Might make it a breaking change, and we should tag if for v2.0. But from a safety perspective, I generally really like it.

**Comment by StLyn4** [comment]
> Hey @StLyn4
> 
> Big thanks for your PR, much appreciated!
> 
> So, I was looking over the changes you suggested about error handling. Ever since we kicked passport-local-mongoose to the curb, it seems like the error messages are behaving pretty well now. Could you take a second look and let me know if you still see any issues? If things are cool, could you roll back the error handling part of your PR?
> 
> The rest of your changes look good! About hiding away the unhandled errors, I'm a bit concerned that hiding these could mess up things for someone who's relying on those error messages. Might make it a breaking change, and we should tag if for v2.0. But from a safety perspective, I generally really like it.

Hi! I completely agree. I noted in the checklist at the very beginning that this could be a breaking change.
I just synced the branch, edited the changes and did a rebase (for a cleaner history).
Everything works, no conflicts, tests passed 👌

To summarize, at the moment the list of changes remains the same, except for the `passport-local-mongoose` section.

**Review by AlessioGr — APPROVED** [review]

**Comment by StLyn4** [comment]
I see that again there is a small conflict in the files because of the master branch update. If you decide to accept my changes closer to the 2.0 release, please ping me when it happens and I will fix the conflicts ASAP 😀

File: test/auth/removed-token/int.spec.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    await client.login();
+  });
+
+  afterAll(async () => {
+    await mongoose.connection.dropDatabase();
+    await mongoose.connection.close();
+    await payload.mongoMemoryServer.stop();
+  });
+
+  it('should not be token in response from /login', async () => {
```
Comment (JarrodMFlesch) [inline test/auth/removed-token/int.spec.ts:28]: Simple english correction.

should not ~~be~~ -> should not include

File: test/auth/removed-token/int.spec.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    const { status, data } = await client.endpoint(`/api/${collectionSlug}/login`, 'post', devUser);
+    expect(status).toBe(200);
+    expect(data.token).not.toBeDefined();
+    expect(data.user.email).toBeDefined();
+  });
+
+  it('should not be token in response from /me', async () => {
```
Comment (JarrodMFlesch) [inline test/auth/removed-token/int.spec.ts:35]: Simple english correction.

should not ~~be~~ -> should not include

File: test/auth/removed-token/int.spec.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    const { status, data } = await client.endpointWithAuth(`/api/${collectionSlug}/me`);
+    expect(status).toBe(200);
+    expect(data.token).not.toBeDefined();
+    expect(data.user.email).toBeDefined();
+  });
+
+  it('should not be token in response from /refresh-token', async () => {
```
Comment (JarrodMFlesch) [inline test/auth/removed-token/int.spec.ts:42]: Simple english correction.

should not ~~be~~ -> should not include

File: test/auth/removed-token/int.spec.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    const { status, data } = await client.endpointWithAuth(`/api/${collectionSlug}/refresh-token`, 'post');
+    expect(status).toBe(200);
+    expect(data.refreshedToken).not.toBeDefined();
+    expect(data.user.email).toBeDefined();
+  });
+
+  it('should not be token in response from /reset-password', async () => {
```
Comment (JarrodMFlesch) [inline test/auth/removed-token/int.spec.ts:49]: Simple english correction.

should not ~~be~~ -> should not include

**Review by JarrodMFlesch — COMMENTED** [review]
Overall I think this looks good. I like these changes/improvements a lot 👍

**Review by DanRibbens — APPROVED** [review]

**Comment by AlessioGr** [comment]
Thank you a lot for your PR again - I've just merged it into 2.0 via #3214! (Was easier to merge it from a separate branch, as there were some weird issues with the commit history when moving from master to 2.0).

**Pull request closed without merging**